Backported from the upstream commit.

From 9c869d08d82c73f62ab3527567858ce4b0cf1257 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Zbigniew=20J=C4=99drzejewski-Szmek?= <zbyszek@in.waw.pl>
Date: Wed, 18 Jul 2018 17:26:17 +0200
Subject: [PATCH] meson: unify linux/stat.h check with other checks and use
 _GNU_SOURCE

Using _GNU_SOURCE is better because that's how we include the headers in the
actual build, and some headers define different stuff when it is defined.
sys/stat.h for example defines 'struct statx' conditionally.

diff -purN systemd/meson.build systemd-izh/meson.build
--- systemd/meson.build	2020-10-13 14:07:06.552552990 +0000
+++ systemd-izh/meson.build	2020-10-15 16:08:55.058347496 +0000
@@ -467,7 +467,6 @@ decl_headers = '''
 #include <uchar.h>
 #include <linux/ethtool.h>
 #include <linux/fib_rules.h>
-#include <linux/stat.h>
 #include <sys/stat.h>
 '''
 # FIXME: key_serial_t is only defined in keyutils.h, this is bound to fail
@@ -481,10 +480,24 @@ foreach decl : ['char16_t',
                ]
 
         # We get -1 if the size cannot be determined
-        have = cc.sizeof(decl, prefix : decl_headers) > 0
+        have = cc.sizeof(decl, prefix : decl_headers, args : '-D_GNU_SOURCE') > 0
+
+        if decl == 'struct statx'
+                if have
+                        want_linux_stat_h = false
+                else
+                        have = cc.sizeof(decl,
+                                         prefix : decl_headers + '#include <linux/stat.h>',
+                                         args : '-D_GNU_SOURCE') > 0
+                        want_linux_stat_h = have
+                endif
+        endif
+
         conf.set10('HAVE_' + decl.underscorify().to_upper(), have)
 endforeach
 
+conf.set10('WANT_LINUX_STAT_H', want_linux_stat_h)
+
 foreach decl : [['IFLA_INET6_ADDR_GEN_MODE',         'linux/if_link.h'],
                 ['IN6_ADDR_GEN_MODE_STABLE_PRIVACY', 'linux/if_link.h'],
                 ['IFLA_VRF_TABLE',                   'linux/if_link.h'],
diff -purN systemd/src/basic/missing.h systemd-izh/src/basic/missing.h
--- systemd/src/basic/missing.h	2020-10-13 10:34:00.118952190 +0000
+++ systemd-izh/src/basic/missing.h	2020-10-15 16:09:39.066538168 +0000
@@ -34,7 +34,11 @@
 #include <linux/neighbour.h>
 #include <linux/oom.h>
 #include <linux/rtnetlink.h>
+
+#if WANT_LINUX_STAT_H
 #include <linux/stat.h>
+#endif
+
 #include <net/ethernet.h>
 #include <stdlib.h>
 #include <sys/resource.h>

